doc/man1/openssl.pod

```diff
@@ -778,11 +778,266 @@ client.
 
 =item B<-xcertform> B<DER>|B<PEM>, B<-xkeyform> B<DER>|B<PEM>
 
+The input format for the extra certifcate and key, respectively.
+See L<openssl(1)/Format Options> for details.
+<<<<<<< dc5af7967d7a4b6afd777a460f0c8918768e256e
+=======
+
+=item B<-xchain_build>
+
+Specify whether the application should build the certificate chain to be
+provided to the server for the extra certificates via the B<-xkey>,
+B<-xcert>, and B<-xchain> options.
+>>>>>>> Move verify options to openssl.pod
+
+=item B<-xcertform> B<DER>|B<PEM>, B<-xkeyform> B<DER>|B<PEM>
+
 The input format for the extra certifcate and key, respectively.
 See L<openssl(1)/Format Options> for details.
 
 =back
 
+=head2 Verification Options
+
+Many OpenSSL commands verify certificates.  The commands try to proceed
+after any validation error, in order to allow all the problems with a
+certificate chain to be determined.
```

Comment:
Is that true for commands other than `verify(1)`?

Reply:
They all *try* to proceed.  Sometimes they don't.  So, yes.  Weasel-wording :)

Reply:
The reason I asked, is that `s_client(1)` and `s_server(1)` have a `-verify_return_error` option, that makes the first error fatal, and the command then returns a non-zero exit code when peer chain verification fails.  They also use a different callback function than the one used by `verify(1)`.  The latter continues only for some errors and not others:
```
        /*
         * Pretend that some errors are ok, so they don't stop further
         * processing of the certificate chain.  Setting ok = 1 does this.
         * After X509_verify_cert() is done, we verify that there were
         * no actual errors, even if the returned value was positive.
         */    
        switch (cert_error) {
        case X509_V_ERR_NO_EXPLICIT_POLICY:
            policies_print(ctx);
            /* fall thru */                                                 
        case X509_V_ERR_CERT_HAS_EXPIRED:
            /* Continue even if the leaf is a self signed cert */
        case X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT:
            /* Continue after extension errors too */
        case X509_V_ERR_INVALID_CA:
        case X509_V_ERR_INVALID_NON_CA:
        case X509_V_ERR_PATH_LENGTH_EXCEEDED:
        case X509_V_ERR_INVALID_PURPOSE:
        case X509_V_ERR_CRL_HAS_EXPIRED:
        case X509_V_ERR_CRL_NOT_YET_VALID:
        case X509_V_ERR_UNHANDLED_CRITICAL_EXTENSION:
            ok = 1;
        }
```
While the callback used by `s_server` and `s_client` continues on almost all errors:
```
    if (!ok) {
        BIO_printf(bio_err, "verify error:num=%d:%s\n", err,
                   X509_verify_cert_error_string(err));
        if (verify_args.depth < 0 || verify_args.depth >= depth) {
            if (!verify_args.return_error)
                ok = 1;
            verify_args.error = err;
        } else {
            ok = 0;
            verify_args.error = X509_V_ERR_CERT_CHAIN_TOO_LONG;
        }
    }
```
except for its own synthetic "chain too long" error, which is a mistake, since I fixed the handling of long chains in libcrypto as of 1.1.0. The special case should simply be removed. This will allow chains one longer than the depth limit argument, since the documented limit bounds the number of intermediate CAs, not the index of the last element.

Reply:
Actually the other apps also have their own custom callbacks:
```text
apps/cms.c:        X509_STORE_set_verify_cb(store, cms_cb);
apps/smime.c:        X509_STORE_set_verify_cb(store, smime_cb);
apps/ts.c:    X509_STORE_set_verify_cb(cert_ctx, verify_cb);
apps/verify.c:    X509_STORE_set_verify_cb(store, cb);
```
In `cms(1)`, `smime(1)` and `ts(1)` the first error is fatal.

Reply:
Thanks for the detailed comments (and research).  I'll address them over the weekend.

Reply:
I fixed the documentation of the commands you noted above to properly explain the behavior on verification error.